src/main_nep/nep4.cu

```diff
@@ -261,18 +264,27 @@ static __global__ void apply_gnn_message_passing(
     // get messages for atom i and neighbors
     float q_theta_i[MAX_DIM] = {0.0f};
     float q_theta_j[MAX_NEIGHBORS * MAX_DIM] = {0.0f}; // maximum size when all atoms are neighbors
+    float fc_ij[MAX_DIM] = {0.0f};
     for (int d = 0; d < annmb.dim; ++d) {
       q_theta_i[d] = g_messages[n1 + d * N];
     }
     for (int j = 0; j < neighbor_number; ++j) {
-      int n2 = g_NL[n1 + N * j];
+      int index = n1 + N*j;
+      int n2 = g_NL[index];
       for (int d = 0; d < annmb.dim; ++d) {
         q_theta_j[j + d * neighbor_number] = g_messages[n2 + d * N];
+        // Compute weight fc_ij
+        float r12[3] = {g_x12[index], g_y12[index], g_z12[index]};
+        // Can these be saved beforehand? Or is it too memory intensive?
+        float d12 = sqrt(r12[0] * r12[0] + r12[1] * r12[1] + r12[2] * r12[2]);
+        float fc12;
+        find_fc(paramb.rc_angular, paramb.rcinv_angular, d12, fc12);
+        fc_ij[j] = fc12;
```

Comment:
calculation of fc_ij can be outside of this loop over d, right after `int n2 = g_NL[index];`

Reply:
Ah yes of course, now I'm basically doing it `N_des` times. I don't seem to have been very much awake this morning :laughing: